### PR TITLE
Feature/allow propagation edition

### LINKED ISSUE #529: Allow propagation edition
## Overview
The way permissions propagate between item ([described in doc](https://france-ioi.github.io/algorea-devdoc/design/access-rights/items/#propagation-across-parent-child-items)) is current not editable. Create a first UI to update it, inspired by [designs](https://france-ioi.github.io/algorea-designs/03.Activities_Edit_05.Edit-chapter-content)  and what was done in the [design branch](http://dev.algorea.org/branch/design) ([code is still available](https://github.com/France-ioi/AlgoreaFrontend/tree/design)).

## Acceptance criteria

- [ ] In item edit mode, in the children list, on the right of each row: add an icon in blue:
  - `fa-eye` if `content_view_propagation` is `as_content` 
  - `fa-lock` if `content_view_propagation` is `as_info` 
  - `fa-eye-slash` if `content_view_propagation` is `none` 
- [ ] when clicking on the icon, a blue overlay is displayed (either using ng-overlaypanel or the same way it was done in the [design branch](https://github.com/France-ioi/AlgoreaFrontend/tree/design)).
- [ ] this panel has 3 options: "locked and hidden" (`none`) , "locked" (`as_info`) , "open" (`as_content`) with the same icon in front
- [ ] on click, the panel closes (shorten/disable animation so that's fast (as in group-progress grid)) and the icon is updated
- [ ] on save, the selected value is saved as `content_view_propagation`
- [ ] depending on the permissions of the current user (in items.permissions), so option may be disabled:
  - `none` is always allowed
  - `as_info` is allowed  if can_grant_view != 'none' or if the current propagation is `as_info` or 'as_content'
  - `as_content` is allowed if the can_grant_view >= 'content' or better, or if the current propagation is 'as_content'

### PR DESCRIPTION
**Pull request by Iloveall**
Fixes #529

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]
Initial feedback just based on testing:
- very good 👍 
- [x] remove the "?" icon as it is not clickable nor showing anything
- [ ] on hover, make the cell highlight in a way. Maybe by making the background a bit darker?
- [ ] we might forget on which line we click to display this overlay. If easy, could you add the triangle "arrow"?
Screenshot 2021-07-20 at 17 07 14
- [ ] it would be interesting to "highlight" the current value among the three options, propose a way to highlight it.
- [ ] "Default access" could be on a different background as it is not a selectable option.
- [ ] When I add a new searched existing item, there is no icon but the menu is still displayed !?
Screenshot 2021-07-20 at 17 12 52. You would need the permissions to know what's possible. While not in the service, just don't let the user change it before a save.
- [ ] same for a newly created item. As a newly created item will belong to current user, you can consider that the user has full access on it (so all options should be available), default to "as_info".

(these last two should have been found by you during your tests (to ask me what to do))

File: src/app/modules/item/components/propagation-edit-menu/propagation-edit-menu.component.html [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      [ngClass]="{'active': childData?.contentViewPropagation === 'as_content'}"
+      (click)="onClick('as_content')"
+      *ngIf="['content', 'content_with_descendants', 'solution', 'solution_with_grant'].includes(childData?.permissions?.canGrantView || '') || childData?.contentViewPropagation === 'as_content'"
+  >
+    <div class="menu-item-title" i18n>Open</div>
+    <div class="menu-item-icon">
+      <i class="fa fa-eye"></i>
+    </div>
+  </li>
+</ul>
```
Comment (smadbe) [inline src/app/modules/item/components/propagation-edit-menu/propagation-edit-menu.component.html:35]: Small bug/glitch: When there is only one option in the "popup" ("locked and hidden") and I click on it, I have very briefly the "locked" option appearing while the popup is disappearing.

**Review by smadbe — CHANGES_REQUESTED** [review]

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts [hunk 2]
```diff
@@ -45,11 +47,15 @@ export const DEFAULT_SCORE_WEIGHT = 1;
...
 
+  @ViewChild('op') op?: OverlayPanel;
+
   state: 'loading' | 'error' | 'ready' = 'ready';
   data: ChildData[] = [];
   selectedRows: ChildData[] = [];
   scoreWeightEnabled = false;
   addedItemIds: string[] = [];
+  propagationEditItemIdx: number | null = null;
+  propagationEditItem?: ChildData | null = null;
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.ts]: Fixed

File: src/app/modules/item/components/propagation-edit-menu/propagation-edit-menu.component.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Component, Output, EventEmitter, Input } from '@angular/core';
+import { ChildData } from '../item-children-edit/item-children-edit.component';
+
+@Component({
+  selector: 'alg-propagation-edit-menu',
+  templateUrl: 'propagation-edit-menu.component.html',
+  styleUrls: [ 'propagation-edit-menu.component.scss' ],
+})
+export class PropagationEditMenuComponent {
+  @Input() childData?: ChildData | null;
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/propagation-edit-menu/propagation-edit-menu.component.ts]: Fixed

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts [hunk 4]
```diff
@@ -45,11 +47,15 @@ export const DEFAULT_SCORE_WEIGHT = 1;
...
   @Input() itemData?: ItemData;
 
+  @ViewChild('op') op?: OverlayPanel;
+
   state: 'loading' | 'error' | 'ready' = 'ready';
   data: ChildData[] = [];
   selectedRows: ChildData[] = [];
   scoreWeightEnabled = false;
   addedItemIds: string[] = [];
+  propagationEditItemIdx: number | null = null;
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.ts]: Fixed

File: src/app/modules/item/components/propagation-edit-menu/propagation-edit-menu.component.html (hunk 1 above)
Comment (Iloveall) [reply src/app/modules/item/components/propagation-edit-menu/propagation-edit-menu.component.html:35]: Fixed

**Review by smadbe — APPROVED** [review]
👍

**Pull request merged**